### PR TITLE
Auth0 middleware

### PR DESCRIPTION
**Pull request by gamebot770**
### Purpose
Hooks in middleware that will allow for Auth0 to authenticate any and all requests sent to hootfs. 

### Details
Authenticated users will have their information stored in the context under the key "userInfo". 
Currently will only hold a string representing the user's email. Can add more information upon request.

### Modifications required to other parts of codebase
No code changes will be required anywhere else in the codebase. Verifier is wrapped in the google_verifer package

### CODE REVIEW

**Comment by Landmaster** [comment]
Shouldn't .idea/workspace.xml be gitignored?

File: .idea/workspace.xml [hunk 1]
```diff
@@ -0,0 +1,136 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (josh8551021) [inline .idea/workspace.xml:1]: nit: remove JetBrains workspace file.

File: src/core/auth/google_verifier.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+	}
+	var result map[string]interface{}
+	err = json.Unmarshal(body, &result)
+
+	if err != nil {
+		log.Println("Could not unmarshal the json: ", err)
+		return false, fmt.Errorf(string(body))
+	}
+
+	//for key, value := range result {
```
Comment (josh8551021) [inline src/core/auth/google_verifier.go:74]: nit: remove this commented code

File: src/core/hootfs/head_node.go [hunk 3]
```diff
@@ -80,12 +83,15 @@ func (fms *HootFsServer) GetDirectoryContentsAsProto(dirId uuid.UUID) ([]*head.O
...
+	log.Println("Starting Server")
+	verifier := google_verifer.New("https://dev-dewy8ew9.us.auth0.com/userinfo")
 	if err != nil {
 		log.Fatalf("Failed to listen: %v", err)
 	}
 	var opts []grpc.ServerOption
+	opts = append(opts, grpc.UnaryInterceptor(grpc_auth.UnaryServerInterceptor(verifier.Authenticate)))
 	s := grpc.NewServer(opts...)
-
+	print("Server started")
```
Comment (josh8551021) [inline src/core/hootfs/head_node.go:94]: Should this be another `log.println()` call?

File: src/core/auth/google_verifier.go [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+	"io/ioutil"
+	"log"
+	"net/http"
+)
+
+type Google_verifier struct {
+	Url string
+}
+
+func New(url string) Google_verifier {
```
Comment (josh8551021) [inline src/core/auth/google_verifier.go:18]: You should probably return a pointer here, as the `Google_verifier` struct could technically grow in size.

File: src/core/auth/google_verifier.go [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+		log.Println("Could not unmarshal the json: ", err)
+		return false, fmt.Errorf(string(body))
+	}
+
+	//for key, value := range result {
+	//	// Each value is an interface{} type, that is type asserted as a string
+	//	fmt.Println(key, value)
+	//}
+
+	if username == result["email"].(string) {
```
Comment (josh8551021) [inline src/core/auth/google_verifier.go:79]: You should probably check for existence of "email" field, just to be safe.

File: src/core/auth/google_verifier.go [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+type Google_verifier struct {
+	Url string
+}
+
+func New(url string) Google_verifier {
+	googleV := Google_verifier{url}
+	return googleV
+}
+
+func (g Google_verifier) Authenticate(ctx context.Context) (context.Context, error) {
```
Comment (josh8551021) [inline src/core/auth/google_verifier.go:23]: Similar to above, you should probably define this function for a `Google_verifier` pointer, same with all other functions.

**Review by josh8551021 — CHANGES_REQUESTED** [review]
Looks good, just a couple of things.

File: README.md [hunk 7]
```diff
@@ -1 +1,7 @@
-# hootfs
\ No newline at end of file
+# hootfs
+
+##Debug Instructions for GoLand
```
Comment (josh8551021) [inline README.md:3]: Add space here so that header shows up properly. Same with line 4.

File: .idea/workspace.xml (hunk 1 above)
Comment (josh8551021) [reply .idea/workspace.xml]: Maybe add `.idea/` to gitignore?

**Review by josh8551021 — APPROVED** [review]
LGTM, just remove the `workspace.xml` file from the repo before merging.

**Review by chathamabate — APPROVED** [review]
LGTM.

**Review by Landmaster — APPROVED** [review]
should be fine now that workspace xml is not in commit

**Pull request merged**